contracts/utils/cryptography/P256.sol

```diff
@@ -0,0 +1,363 @@
+     * expensive to use, and to deploy, than the version that rely on precompiles.
+     *
+     * @param h - hashed message
+     * @param r - signature half R
+     * @param s - signature half S
+     * @param qx - public key coordinate X
+     * @param qy - public key coordinate Y
+     */
+    function verifySolidity(uint256 h, uint256 r, uint256 s, uint256 qx, uint256 qy) internal view returns (bool) {
+        if (r == 0 || r >= N || s == 0 || s >= N || !isOnCurve(qx, qy)) return false;
+
+        JPoint[16] memory points = _preComputeJacobianPoints(qx, qy);
+        uint256 w = Math.invModPrime(s, N);
+        uint256 u1 = mulmod(h, w, N);
+        uint256 u2 = mulmod(r, w, N);
+        (uint256 x, ) = _jMultShamir(points, u1, u2);
+        return (x == r);
+    }
+
+    /**
+     * @dev public key recovery
+     * @param h - hashed message
+     * @param v - signature recovery param
+     * @param r - signature half R
+     * @param s - signature half S
+     */
+    function recovery(uint256 h, uint8 v, uint256 r, uint256 s) internal view returns (uint256, uint256) {
+        if (r == 0 || r >= N || s == 0 || s >= N || v > 1) return (0, 0);
+
+        uint256 rx = r;
+        uint256 ry2 = addmod(mulmod(addmod(mulmod(rx, rx, P), A, P), rx, P), B, P); // weierstrass equation y² = x³ + a.x + b
+        uint256 ry = Math.modExp(ry2, P1DIV4, P); // This formula for sqrt work because P ≡ 3 (mod 4)
+        if (mulmod(ry, ry, P) != ry2) return (0, 0); // Sanity check
+        if (ry % 2 != v % 2) ry = P - ry;
+
+        JPoint[16] memory points = _preComputeJacobianPoints(rx, ry);
+        uint256 w = Math.invModPrime(r, N);
+        uint256 u1 = mulmod(N - (h % N), w, N);
+        uint256 u2 = mulmod(s, w, N);
+        (uint256 x, uint256 y) = _jMultShamir(points, u1, u2);
+        return (x, y);
+    }
+
+    /**
+     * @dev address recovery
+     * @param h - hashed message
+     * @param v - signature recovery param
+     * @param r - signature half R
+     * @param s - signature half S
+     */
+    function recoveryAddress(uint256 h, uint8 v, uint256 r, uint256 s) internal view returns (address) {
+        (uint256 qx, uint256 qy) = recovery(h, v, r, s);
+        return getAddress(qx, qy);
+    }
+
+    /**
+     * @dev derivate public key
+     * @param privateKey - private key
+     */
+    function getPublicKey(uint256 privateKey) internal view returns (uint256, uint256) {
+        (uint256 x, uint256 y, uint256 z) = _jMult(GX, GY, 1, privateKey);
+        return _affineFromJacobian(x, y, z);
+    }
+
+    /**
+     * @dev Hash public key into an address
+     * @param qx - public key coordinate X
+     * @param qy - public key coordinate Y
+     */
+    function getAddress(uint256 qx, uint256 qy) internal pure returns (address result) {
+        assembly ("memory-safe") {
+            mstore(0x00, qx)
+            mstore(0x20, qy)
+            result := keccak256(0x00, 0x40)
+        }
+    }
```

Comment:
> Using a bytes32 as a "public key identifier" feels more natural to me

That "preverts" us from using AccessControl to give roles to the keys ... but I see your point.

Lets remove them, and we can always re-add them latter if we want

Reply:
> That "preverts" us from using AccessControl to give roles to the keys

Right, this is a good point. I guess it remains to be seen whether users would do this or assign the role to a controlled account using 1271 in the end.

Agree, let's solve it for now and we can discuss it better later.

Reply:
We already discussed that ...

If you use an ERC-1271 account you are just moving who does the verification. The ERC-1271 contract will get a hash and a signature ... it will recover a public key of the signer ... and then what ? It needs to check that this identity is an authorized signer. This is where the ERC-1271 contract could use AccessControl ... but only if it turns the public key it just recovered into an address

Reply:
At some point, someone must check "is that P256 key that I just verified an authorized identity". 

That is when it's useful to represent it as an address

Reply:
> and then what ? It needs to check that this identity is an authorized signer. This is where the ERC-1271 contract could use AccessControl ... but only if it turns the public key it just recovered into an address

This sounds like we haven't discussed it enough.
Regardless, I don't think it makes sense to discuss right now since we don't have evidence that users would like to use ERC-1271 over non-accessible addresses.

> but only if it turns the public key it just recovered into an address

Not really. A contract can be just a signer defined by an internal public key:

```solidity
contract P256Identity is IERC1271 {
    using P256 for bytes32;

    // bytes4(keccak256("isValidSignature(bytes,bytes)")
    bytes4 internal constant EIP1271_MAGIC_VALUE = 0x20c13b0b;

    struct PublicKey {
        bytes32 qx;
        bytes32 qy;
    }

    PublicKey publicKey;

    constructor(bytes32 qx, bytes32 qy) {
        publicKey = PublicKey(qx, qy);
    }

    function isValidSignature(
        bytes32 digest,
        bytes memory signature
    ) external view returns (bytes4) {
        return
            _verifyP256Identity(digest, signature)
                ? EIP1271_MAGIC_VALUE
                : bytes4(0);
    }

    function _verifyP256Identity(
        bytes32 digest,
        bytes memory signature
    ) internal view returns (bool) {
        PublicKey memory pubKey = publicKey();

        (uint256 r, uint256 s) = _splitSignature(signature);

        if (normalize) {
            digest = sha256(abi.encode(digest));
        }

        return digest.verify(digest, r, s, pubKey.qx, pubKey.qy);
    }
    function _splitSignature(
        bytes memory signature
    ) private pure returns (uint256 r, uint256 s) {
        r = uint256(bytes32(signature[00:31]));
        s = uint256(bytes32(signature[32:63]));
    }
}
```

In this way, it's clear that the contract has an address, is controlled by a single private key, and it's already composable with the broader ecosystem. Also it doesn't allow to deposit ETH that might get lost.